File: javascript/strings/AB.js

//Solution to top coder problem 
//Generate a string comprising of A and B 
//Input will be n,k where n is the length of string and k is a 
//number pair in the string of length n where str[i] !== str[j]
function generateRanges(n) {
    var i = 1,
        j = n - 1,
        arr = [];
    while (i < n / 2 && j >= n / 2) {
        arr.push(i * j);
        i++;
        j--;
    }
    return arr;
}

function generateString(n, k) {
    var ranges = generateRanges(n),
        numberOfBInString = [],
        i = 0,
        str = '';

    if ( k > (n * n-1 )/4) {
		return '';
	}

    numberOfBInString = k < n ? [1, n - k] : binarySearch(ranges, k);
	j = 0;
	

	for (i = 0 ; i < n - numberOfBInString[0] ;i++) {
		if (i === n - numberOfBInString[1] - numberOfBInString[0]) {
			str += 'b';	
		}
		str += 'a';
	}
	i = str.length;
	while ( i < n) {
		str +='b';
		i++;
	}
	
    return str;
}

function binarySearch(arr, k) {
    var first = 0,
        last = arr.length - 1;

    while (first >= 0 && last < arr.length) {
        mid = (last - first) / 2;
        if (k > arr[mid]) {
            if (k < arr[mid + 1]) {
                return [mid + 2, arr[mid + 1] - k];
            }
            first = mid + 1;
        } else if (k < arr[mid]) {
            if (k > arr[mid - 1]) {
                return [mid+1, arr[mid] - k];
            }
            last = mid - 1;
        } else {
		     return [mid, arr[mid] - k];
        }
    }
}

//var str1 = generateString(7,2);
var str2 = generateString(7,8);
var str3 = generateString(7,11);
var str4 = generateString(7,13);

File: python/AB.py
import time
class AB:
	def generateRanges(self,n):
		i  = 1
		j = n - 1
		arr = []
		while i <= j:
			arr.append( i * j)
			i = i + 1
			j = j -1
		print arr
		return arr


	def createString(self, n, k ):
		numberOfBInString = []
		str = ''
		j = 0
		i = 0
		
		if n < 2 or n > 50 or k > (n * n+1)/2:
			return ''

		if k == 0:
			return 'A'*n

		ranges = self.generateRanges(n)	
		print len(ranges)
		
		if len(ranges) == 1:
			return 'A'* (n-1) + 'B'

		if k < n:
			numberOfBInString = [1, n - k]
		else:
			numberOfBInString = self.linearSearch(ranges, k) 	

		if numberOfBInString is None:
			return ''

		while i < n - numberOfBInString[0]:
			if i == n - numberOfBInString[1] - numberOfBInString[0]:
				str  = str + 'B'
			str = str + 'A'
			i = i + 1

		slen = len(str)
		
		while slen < n:
			str  = str + 'B'
			slen = slen + 1

		return str


	def linearSearch(self, ranges, k):
		
		if len(ranges) >= 2:
			i = 1
			if k > ranges[i-1]  and k <= ranges[i]:
				print [i+1, ranges[i] - k]
				return [i+1, ranges[i] - k]


start_time = time.time()
ab = AB()
print ab.createString(3,2)
print("--- %s seconds ---" % (time.time() - start_time))

#print generateString(10, 12)